File: frontend/src/lib/api/pdfs.ts
import { requestBackend } from "./client";
import type { PackId } from "../../types/billing";

export interface PdfBalance {
  total: number;
  used: number;
  remaining: number;
}

export function pdfsFromErrorBody(body: unknown): PdfBalance | null {
  if (!body || typeof body !== "object" || !("detail" in body)) return null;
  const detail = (body as { detail?: { pdfs?: PdfBalance } }).detail;
  const pdfs = detail?.pdfs;
  if (
    !pdfs ||
    typeof pdfs.total !== "number" ||
    typeof pdfs.used !== "number"
  ) {
    return null;
  }
  return pdfs;
}

export function getPdfSaves() {
  return requestBackend<PdfBalance>("/api/pdfs");
}

export function grantPdfSaves(packId: PackId) {
  return requestBackend<PdfBalance>("/api/pdfs/grant", {
    method: "POST",
    body: { pack_id: packId },
  });
}

export function consumePdfSave() {
  return requestBackend<PdfBalance & { consumed: boolean }>("/api/pdfs/consume", {
    method: "POST",
  });
}

export function refundPdfSave() {
  return requestBackend<PdfBalance>("/api/pdfs/refund", {
    method: "POST",
  });
}

File: backend/app/routers/pdfs.py
from fastapi import APIRouter, Depends, HTTPException, status

from app.auth.supabase_jwt import CurrentUser, get_current_user
from app.schemas.pdfs import ConsumePdfResult, GrantPdfsRequest, PdfBalance
from app.services.pdfs import (
    consume_result,
    consume_save,
    get_balance,
    grant_for_pack,
    refund_save,
)

router = APIRouter(prefix="/api/pdfs", tags=["pdfs"])


@router.get("", response_model=PdfBalance)
def read_pdfs(user: CurrentUser = Depends(get_current_user)) -> PdfBalance:
    return get_balance(user.id)


@router.post("/grant", response_model=PdfBalance)
def grant_pdfs(
    body: GrantPdfsRequest,
    user: CurrentUser = Depends(get_current_user),
) -> PdfBalance:
    return grant_for_pack(user.id, body.pack_id)


@router.post("/consume", response_model=ConsumePdfResult)
def consume_pdf(user: CurrentUser = Depends(get_current_user)) -> ConsumePdfResult:
    consumed, balance = consume_save(user.id)
    if not consumed:
        raise HTTPException(
            status_code=status.HTTP_402_PAYMENT_REQUIRED,
            detail={
                "message": "No PDF saves remaining",
                "pdfs": balance.model_dump(),
            },
        )
    return consume_result(True, balance)


@router.post("/refund", response_model=PdfBalance)
def refund_pdf(user: CurrentUser = Depends(get_current_user)) -> PdfBalance:
    return refund_save(user.id)
